### PR TITLE
Fix/command out err

### LINKED ISSUE #287: bug in BuildTestCommand, only retrieving first line
Here is an example of BuildTestCommand, its retrieving the first line of output which is wrong.
This affects the content of output and error stream written to file.
```
>>> from buildtest.utils.command import BuildTestCommand
>>> cmd = BuildTestCommand("python .buildtest/python-circle/circle_area.py")
>>> out,err = cmd.execute()
>>> out
['Circle Radius  2\n']

>>> os.system("python .buildtest/python-circle/circle_area.py")
Circle Radius  2
Area of circle  12.566370614359172
Block
0
```

### PR DESCRIPTION
**Pull request by vsoch**
I'm unable to reproduce this error, but logically based on the general execution envronment that @shahzebsiddiqui is describing, I suspect there is an issue with identifying the newline character and so this is the first thing I'd want to try for the fix.

> When stdout or stderr are pipes and universal_newlines is True then all line endings will be converted to '\n' as described for the universal newlines 'U' mode argument to open().

The fix is not to revert back to a less secure method, but to properly debug the actual issue. This will close #287

### CODE REVIEW

**Comment by shahzebsiddiqui** [comment]
@vsoch I am not sure what i am to debug, is this PR still in development, should i wait until you are done?

**Comment by vsoch** [comment]
It’s done. I figured you’d want to test it to see if it works for you.

**Comment by shahzebsiddiqui** [comment]
Okay so i tested this further and results are not consistent at all, some commands are working while others are not and you can test these also. I got ``env``  and ``gcc --version`` to spit out the right results.
But running ``python --help``, ``lscpu``. ``cat /etc/profile`` resulted in first line.

SUCESS
---------

```
>>> cmd = BuildTestCommand("env")
>>> cmd.execute()
(['SHELL=/bin/bash\n', ':"local","kind":"builtin"},"vscode.@theia/vscode-builtin-markdown@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-markdown@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-npm@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-npm@0.2.1","url":"local","kind":"builtin"},"vscode.objective-c@1.39.1":{"fullPluginName":"vscode.objective-c@1.39.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-perl@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-perl@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-php@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-php@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-powershell@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-powershell@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-pug@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-pug@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-python@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-python@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-r@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-r@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-razor@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-razor@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-ruby@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-ruby@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-rust@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-rust@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-scss@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-scss@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-shaderlab@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-shaderlab@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-shellscript@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-shellscript@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-sql@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-sql@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-swift@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-swift@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-typescript@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-typescript@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-typescript-language-features@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-typescript-language-features@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-vb@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-vb@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-xml@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-xml@0.2.1","url":"local","kind":"builtin"},"vscode.@theia/vscode-builtin-yaml@0.2.1":{"fullPluginName":"vscode.@theia/vscode-builtin-yaml@0.2.1","url":"local","kind":"builtin"},"redhat.java@0.58.0":{"fullPluginName":"redhat.java@0.58.0","url":"local","kind":"builtin"},"vscjava.vscode-java-debug@0.25.1":{"fullPluginName":"vscjava.vscode-java-debug@0.25.1","url":"local","kind":"builtin"},"vscjava.vscode-java-dependency@0.9.0":{"fullPluginName":"vscjava.vscode-java-dependency@0.9.0","url":"local","kind":"builtin"},"ms-vscode.node-debug@1.38.4":{"fullPluginName":"ms-vscode.node-debug@1.38.4","url":"local","kind":"builtin"},"ms-vscode.node-debug2@1.33.0":{"fullPluginName":"ms-vscode.node-debug2@1.33.0","url":"local","kind":"builtin"},"ms-python.python@2019.11.50794":{"fullPluginName":"ms-python.python@2019.11.50794","url":"local","kind":"builtin"},"ms-vscode.Go@0.11.4":{"fullPluginName":"ms-vscode.Go@0.11.4","url":"local","kind":"builtin"},"redhat.vscode-xml@0.8.0":{"fullPluginName":"redhat.vscode-xml@0.8.0","url":"local","kind":"builtin"},"redhat.vscode-yaml@0.5.2":{"fullPluginName":"redhat.vscode-yaml@0.5.2","url":"local","kind":"builtin"},"felixfbecker.php-intellisense@2.3.10":{"fullPluginName":"felixfbecker.php-intellisense@2.3.10","url":"local","kind":"builtin"},"felixfbecker.php-debug@1.13.0":{"fullPluginName":"felixfbecker.php-debug@1.13.0","url":"local","kind":"builtin"},"castwide.solargraph@0.21.1":{"fullPluginName":"castwide.solargraph@0.21.1","url":"local","kind":"builtin"},"rust-lang.rust@0.7.0":{"fullPluginName":"rust-lang.rust@0.7.0","url":"local","kind":"builtin"},"vscode.theme-abyss@1.39.1-prel":{"fullPluginName":"vscode.theme-abyss@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-kimbie-dark@1.39.1-prel":{"fullPluginName":"vscode.theme-kimbie-dark@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-monokai@1.39.1-prel":{"fullPluginName":"vscode.theme-monokai@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-monokai-dimmed@1.39.1-prel":{"fullPluginName":"vscode.theme-monokai-dimmed@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-quietlight@1.39.1-prel":{"fullPluginName":"vscode.theme-quietlight@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-red@1.39.1-prel":{"fullPluginName":"vscode.theme-red@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-solarized-dark@1.39.1-prel":{"fullPluginName":"vscode.theme-solarized-dark@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-solarized-light@1.39.1-prel":{"fullPluginName":"vscode.theme-solarized-light@1.39.1-prel","url":"local","kind":"builtin"},"vscode.theme-tomorrow-night-blue@1.39.1-prel":{"fullPluginName":"vscode.theme-tomorrow-night-blue@1.39.1-prel","url":"local","kind":"builtin"},"vscode.vscode-theme-seti@1.39.1-prel":{"fullPluginName":"vscode.vscode-theme-seti@1.39.1-prel","url":"local","kind":"builtin"},"vscode.merge-conflict@1.39.1-prel":{"fullPluginName":"vscode.merge-conflict@1.39.1-prel","url":"local","kind":"builtin"},"ms-vscode.references-view@0.0.47":{"fullPluginName":"ms-vscode.references-view@0.0.47","url":"local","kind":"builtin"},"EditorConfig.EditorConfig@0.14.4":{"fullPluginName":"EditorConfig.EditorConfig@0.14.4","url":"local","kind":"builtin"}}\n', 'VSCODE_NLS_CONFIG={"locale":"en","availableLanguages":{}}\n', 'NODE_VERSION=12.16.2\n', 'GITPOD_GIT_USER_EMAIL=shahzebmsiddiqui@gmail.com\n', 'IRBRC=/home/gitpod/.rvm/rubies/ruby-2.6.5/.irbrc\n', 'rvm_path=/home/gitpod/.rvm\n', 'OLDPWD=/home/gitpod\n', 'GOPATH=/workspace/go\n', 'GITPOD_TASKS=[]\n', 'LINES=17\n', 'COLUMNS=165\n'], [])
```

```
>>> cmd = BuildTestCommand("gcc --version")
>>> cmd.execute()
(['gcc (Ubuntu 9.3.0-10ubuntu1) 9.3.0\n', 'Copyright (C) 2019 Free Software Foundation, Inc.\n', 'This is free software; see the source for copying conditions.  There is NO\n', 'warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n'], [])
```

Error
-------

```
>>> cmd = BuildTestCommand("python --help")
>>> cmd.execute()
(['usage: /home/gitpod/.pyenv/versions/3.8.2/bin/python [option] ... [-c cmd | -m mod | file | -] [arg] ...\n'], [])
```

```
Type "help", "copyright", "credits" or "license" for more information.
>>> from buildtest.utils.command import BuildTestCommand
>>> cmd = BuildTestCommand("lscpu")
>>> cmd.execute()
(['Architecture:                    x86_64\n'], [])
```

```
>>> cmd = BuildTestCommand("cat /etc/profile")
>>> cmd.execute()
(['# /etc/profile: system-wide .profile file for the Bourne shell (sh(1))\n'], [])
```

The output of ``lscpu`` is more than one line and it's just getting the first line and stop.

```
gitpod /workspace/buildtest-framework $ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           79
Model name:                      Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                        0
CPU MHz:                         2200.000
BogoMIPS:                        4400.00
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       256 KiB
L1i cache:                       256 KiB
L2 cache:                        2 MiB
L3 cache:                        55 MiB
NUMA node0 CPU(s):               0-15
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm
                                  constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe 
                                 popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bm
                                 i1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
```

**Comment by vsoch** [comment]
No worries, give the new changes a shot.

```python
from buildtest.utils.command import *                                                                                                                          

cmd = BuildTestCommand("python --help")                                                                                                                        
cmd.execute()                                                                                                                                                  
(['usage: /home/vanessa/anaconda3/bin/python [option] ... [-c cmd | -m mod | file | -] [arg] ...\n',
  'Options and arguments (and corresponding environment variables):\n',
  '-b     : issue warnings about str(bytes_instance), str(bytearray_instance)\n',
  '         and comparing bytes/bytearray with str. (-bb: issue errors)\n',
  "-B     : don't write .pyc files on import; also PYTHONDONTWRITEBYTECODE=x\n",
  '-c cmd : program passed in as string (terminates option list)\n',
  '-d     : debug output from parser; also PYTHONDEBUG=x\n',
  '-E     : ignore PYTHON* environment variables (such as PYTHONPATH)\n',
  '-h     : print this help message and exit (also --help)\n',
  '-i     : inspect interactively after running script; forces a prompt even\n',
  '         if stdin does not appear to be a terminal; also PYTHONINSPECT=x\n',
  "-I     : isolate Python from the user's environment (implies -E and -s)\n",
  '-m mod : run library module as a script (terminates option list)\n',
  '-O     : remove assert and __debug__-dependent statements; add .opt-1 before\n',
  '         .pyc extension; also PYTHONOPTIMIZE=x\n',
  '-OO    : do -O changes and also discard docstrings; add .opt-2 before\n',
  '         .pyc extension\n',
  "-q     : don't print version and copyright messages on interactive startup\n",
  "-s     : don't add user site directory to sys.path; also PYTHONNOUSERSITE\n",
  "-S     : don't imply 'import site' on initialization\n",
  '-u     : force the stdout and stderr streams to be unbuffered;\n',
  '         this option has no effect on stdin; also PYTHONUNBUFFERED=x\n',
  '-v     : verbose (trace import statements); also PYTHONVERBOSE=x\n',
  '         can be supplied multiple times to increase verbosity\n',
  '-V     : print the Python version number and exit (also --version)\n',
  '         when given twice, print more information about the build\n',
  '-W arg : warning control; arg is action:message:category:module:lineno\n',
  '         also PYTHONWARNINGS=arg\n',
  '-x     : skip first line of source, allowing use of non-Unix forms of #!cmd\n',
  '-X opt : set implementation-specific option\n',
  '--check-hash-based-pycs always|default|never:\n',
  '    control how Python invalidates hash-based .pyc files\n',
  'file   : program read from script file\n',
  '-      : program read from stdin (default; interactive mode if a tty)\n',
  'arg ...: arguments passed to program in sys.argv[1:]\n',
  'Other environment variables:\n',
  'PYTHONSTARTUP: file executed on interactive startup (no default)\n',
  "PYTHONPATH   : ':'-separated list of directories prefixed to the\n",
  '               default module search path.  The result is sys.path.\n',
  'PYTHONHOME   : alternate <prefix> directory (or <prefix>:<exec_prefix>).\n',
  '               The default module search path uses <prefix>/lib/pythonX.X.\n',
  "PYTHONCASEOK : ignore case in 'import' statements (Windows).\n",
  'PYTHONIOENCODING: Encoding[:errors] used for stdin/stdout/stderr.\n',
  'PYTHONFAULTHANDLER: dump the Python traceback on fatal errors.\n',
  "PYTHONHASHSEED: if this variable is set to 'random', a random value is used\n",
  '   to seed the hashes of str, bytes and datetime objects.  It can also be\n',
  '   set to an integer in the range [0,4294967295] to get hash values with a\n',
  '   predictable seed.\n',
  'PYTHONMALLOC: set the Python memory allocators and/or install debug hooks\n',
  '   on Python memory allocators. Use PYTHONMALLOC=debug to install debug\n',
  '   hooks.\n',
  'PYTHONCOERCECLOCALE: if this variable is set to 0, it disables the locale\n',
  '   coercion behavior. Use PYTHONCOERCECLOCALE=warn to request display of\n',
  '   locale coercion and locale compatibility warnings on stderr.\n',
  'PYTHONBREAKPOINT: if this variable is set to 0, it disables the default\n',
  '   debugger. It can be set to the callable of your debugger of choice.\n',
  'PYTHONDEVMODE: enable the development mode.\n'],
 [])
```

File: buildtest/utils/command.py [hunk 1]
```diff
@@ -54,27 +116,26 @@ def execute(self):
...
 
+            # Iterate through the output
+            while returncode is None:
+                returncode = process.poll()
+
         # Get the remainder of lines, add return code
-        out, err = process.communicate()
-        self.out += ["%s\n" % x for x in self.decode(out).split("\n") if x]
-        self.err += ["%s\n" % x for x in self.decode(err).split("\n") if x]
+        self.out += ["%s\n" % x for x in self.decode(capture.out).split("\n") if x]
```
Comment (shahzebsiddiqui) [inline buildtest/utils/command.py:134]: The output from ``capture.out`` is in string, I am wondering if it makes sense to return output as a ``str`` as pose to ``list`` when we return from method. 

I see the ``self.decode`` is going through each line and decode based on encoding, i am fine with this but maybe once ``self.out`` is a list to convert it back to a string. I can see when invoking ``BuildTestCommand`` to get the output and error and having to convert list to string everytime would be a bit annoying.

File: buildtest/utils/command.py [hunk 2]
```diff
@@ -1,7 +1,69 @@
...
+
+class Capturing:
+    """capture output from stdout and stderr into capture object.
+       This is based off of github.com/vsoch/gridtest but modified
+       to write files. The stderr and stdout are set to temporary files at
+       the init of the capture, and then they are closed when we exit. This
+       means expected usage looks like:
+
+       with Capturing() as capture:
+           process = subprocess.POpen(...)
```
Comment (shahzebsiddiqui) [inline buildtest/utils/command.py]: typo **POpen** --> **Popen**

Comment (shahzebsiddiqui) [reply buildtest/utils/command.py]: It may be nice to add full example in docstring of how to use ``subprocess.Popen`` with Capture class in docstring as pose to **...**

File: buildtest/utils/command.py [hunk 3]
```diff
@@ -1,7 +1,69 @@
...
+        return self
+
+    def set_stdout(self):
+        self.stdout = open(tempfile.mkstemp()[1], "w")
+
+    def set_stderr(self):
+        self.stderr = open(tempfile.mkstemp()[1], "w")
+
+    def __exit__(self, *args):
+        self.stderr.close()
```
Comment (shahzebsiddiqui) [inline buildtest/utils/command.py:42]: I don't think it's possible to run into issue closing file handle but does it make sense to put an exception let's say with **try except**. Only thing i can think of is if file is read and deleted in time before file is closed it could result in an exception.

**Review by shahzebsiddiqui — COMMENTED** [review]
Overall i tested this and it looks promising. I was able to get output for results from running  all the previous shell commands. There are few comments and some edge cases i realized we can't support for instance and maybe this is not intended but we cant run any shell commands such as ``echo $HOME`` or ``ls $HOME`` using BuildTestCommand since we are not using ``shell=True``   in subprocess. This is true with input and output redirection and other shell specific syntax.

File: buildtest/utils/command.py (hunk 1 above)
Comment (vsoch) [reply buildtest/utils/command.py:134]: It's no longer reading from stdout / stderr, but just reading the written file, so probably the decode isn't even necessary. This is also why the custom parsing (e.g., splitting by newline) is handled by the calling class - someone else using the Capturing() class might just want the entire blob of text read it, sans any additional processing. This is why I made this explicit choice.

File: buildtest/utils/command.py (hunk 2 above)
Comment (vsoch) [reply buildtest/utils/command.py]: Sure, feel free to add this after if you like!

Comment (vsoch) [reply buildtest/utils/command.py]: Popen typo fixed!

File: buildtest/utils/command.py (hunk 3 above)
Comment (vsoch) [reply buildtest/utils/command.py:42]: This wouldn't be an error (I'm not sure you tested this?) there isn't any error if the file is closed (and then is closed again)
```
with open('/tmp/test.x', 'w') as filey: 
    print(filey) 

<_io.TextIOWrapper name='/tmp/test.x' mode='w' encoding='UTF-8'>

filey.close()                                                                                        
filey.close()   
```

OR if the file is deleted (and then closed again)

```python
import os
os.remove(filey.name)                                                                               
filey.name                                                                                          
 '/tmp/test.x'
filey.close()                                                                                       
```
Tada, no errors thrown! Good thought though. :)

Comment (shahzebsiddiqui) [reply buildtest/utils/command.py:42]: Nope i didn't test i was just thinking of an edge case where we run into an exception. Thanks for the test, it's good to know we close file-handle gracefully in the event if file is not present.

File: buildtest/utils/command.py (hunk 1 above)
Comment (shahzebsiddiqui) [reply buildtest/utils/command.py:134]: Yea i can myself preferring to read output as a text blob as pose to converting the list to string. Maybe just a thought if we can have a method in ``buildtest/utils/file.py`` let's say **list_to_string** that automatically converts a list to string. It can be as simple as 

```
def list_to_string(content):
  return ' '.join(content)
```

If we add this method i'd be fine with having BuildTestCommand return output as a list so i can convert it to a string.

Reason is i can see BuildTestCommand used in many places where we need output as a string. One example is in https://github.com/buildtesters/buildtest/blob/devel/buildtest/docgen/main.py  which uses BuildTestCommand to run examples and write output to a txt file. This helps create the output of the test automatically for documentation as pose to running each command manually. I found that as buildtest changes its behavior i'd like a script to rebuild all the doc examples.

File: buildtest/utils/command.py (hunk 2 above)
Comment (shahzebsiddiqui) [reply buildtest/utils/command.py]: Thanks

File: buildtest/utils/command.py (hunk 1 above)
Comment (vsoch) [reply buildtest/utils/command.py:134]: You could, but since this function is just one line, and abstracts away a relatively simple operation `content.split("\n")` I don't think this is a useful feature to add. the "split" function provided by the string class already handles this, and in one simple, clear command. The same is the case for join. :-1: from me.

File: buildtest/utils/command.py [hunk 4]
```diff
@@ -1,7 +1,69 @@
+from buildtest.utils.file import read_file
```
Comment (shahzebsiddiqui) [inline buildtest/utils/command.py]: nit pick: can we have ``from buildtest.utils.file import read_file``  after ``imports``

**Comment by shahzebsiddiqui** [comment]
Looks like black failed as soon as checks pass we are good to merge.

**Comment by vsoch** [comment]
I'm on a different computer so I don't have the repo handy and fixed the import order in a text editor, didn't take account for black requiring two newlines. Should be fixed now.

**Comment by shahzebsiddiqui** [comment]
As soon as it's all green good to merge. 👍
